src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Sockets;
+using OpenTelemetry.Instrumentation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Implementation
+{
+    internal class ElasticsearchRequestPipelineDiagnosticListener : ListenerHandler
+    {
+        private readonly ActivitySourceAdapter activitySource;
+        private readonly ElasticsearchClientInstrumentationOptions options;
+        private readonly PropertyFetcher<Uri> uriFetcher = new PropertyFetcher<Uri>("Uri");
+        private readonly PropertyFetcher<object> methodFetcher = new PropertyFetcher<object>("Method");
+        private readonly PropertyFetcher<bool> successFetcher = new PropertyFetcher<bool>("SuccessOrKnownError");
+        private readonly PropertyFetcher<string> debugInformationFetcher = new PropertyFetcher<string>("DebugInformation");
+        private readonly PropertyFetcher<int?> httpStatusFetcher = new PropertyFetcher<int?>("HttpStatusCode");
+        private readonly PropertyFetcher<Exception> orginalExceptionFetcher = new PropertyFetcher<Exception>("OriginalException");
+
+        public ElasticsearchRequestPipelineDiagnosticListener(ActivitySourceAdapter activitySource, ElasticsearchClientInstrumentationOptions options)
+            : base("Elasticsearch.Net.RequestPipeline")
+        {
+            if (activitySource == null)
+            {
+                throw new ArgumentNullException(nameof(activitySource));
+            }
+
+            this.activitySource = activitySource;
+            this.options = options;
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            var uri = this.uriFetcher.Fetch(payload);
+
+            if (uri == null)
+            {
+                ElasticsearchInstrumentationEventSource.Log.NullPayload(nameof(ElasticsearchRequestPipelineDiagnosticListener), nameof(this.OnStartActivity));
+                return;
+            }
+
+            var method = this.methodFetcher.Fetch(payload)?.ToString();
+            var elasticType = uri.Segments.FirstOrDefault(s => !s.Equals("/") && !s.StartsWith("_"));
+            if (elasticType != null && elasticType.EndsWith("/"))
+            {
+                elasticType = elasticType.Substring(0, elasticType.Length - 1);
+            }
+
+            if (!string.IsNullOrEmpty(elasticType))
+            {
+                activity.DisplayName = $"Elasticsearch {method} {elasticType}";
+            }
+            else if (activity.OperationName == "Ping")
+            {
+                activity.DisplayName = $"Elasticsearch Ping";
+            }
+            else
+            {
+                activity.DisplayName = $"Elasticsearch {method}";
```

Comment:
considering the fact that DisplayName is not readily available, I'd suggest to set it *after* activitySource.Start, and inside `if (activity.IsAllDataRequested)`

Reply:
Does that mean I would move the code to get the URI and the method inside of the `if (activity.IsAllDataRequested)`? Would I set any information about the activity outside of the `IsAllDataRequested`?

Reply:
There is no right answer to this.
My take is this:
If the information is readily available, set it.
If it is not readily available, and requires some work, then wait for samplers to run (.Start internally calls sampler). Based on sampling decision, populate more things.

(The more information we set earlier, the more information samplers have to make decision. But then it somewhat defeats the purpose of sampling - which is to save perf. So instrumentations can do a balanced approach.)

DisplayName is something which is allowed to be changed, and because of this, Samplers are not likely to make a decision based on the DisplayName. This is the reason why I recommend setting DisplayName after sampling decision is made.

Reply:
Ok, that makes sense. So I guess I would want to at least get the URI and set some tags from that. Just not the Name and any other additional nice to have info.

Reply:
Well, I'm looking at other instrumentations like the HTTP one and the GRPC one and they are pretty much just calculating and setting the DisplayName outside of the IsDataRequested check and nothing else. Now I'm confused. :-(

Reply:
@ejsmith I think setting DisplayName before calling Start was something I started doing really early on and what you see is just consistency with the early integrations. I don't have a really great reason for doing that, it just felt like the right thing to do at the time. We might have been the first people anywhere to try and do something practical with the ActivitySource API so, TBH, some of the stuff was just made up as we went along 😄  Even now we're still debating things like whether or not we should add the raw objects automatically.

Generally, try to default to the highest performance option available to you.

```csharp
activity.DisplayName = $"Elasticsearch: {method} {uri.AbsolutePath}";
```

Right now it's doing a string allocation which is a kind of notorious performance killer. If that DisplayName was coming from a cache, or was constant, I don't think @cijothomas would have commented on it. Because it is doing a string allocation, if the sampler doesn't sample the Activity, then that is wasted perf.

PS: I think AbsolutePath is good, but can you confirm it is [low cardinality](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span) as the spec defines it?

Reply:
Ok this is making more sense to me now. I am going to change it so that it still sets the displayname but uses a cache. Also the method combined with the absolute path is basically the equivalent of a DB table name and the type of operation SELECT/UPDATE/DELETE. So I think that is probably a pretty good choice of still being useful without being high cardinality. Thanks for explaining!

Reply:
Actually, I'm wrong. If I get a document by id then the document id is included in the path. I will fix that.

Reply:
Ok, I believe this is addressed now. I am defaulting to just using method which is cached before the IsAllDataRequested check and then I am adding the index name if it is available. Also added several new tests.